src/libraries/Microsoft.Bcl.Cryptography/src/System/Security/Cryptography/Shake256.cs

```diff
@@ -4,35 +4,196 @@
+        {
+            int count = Interlocked.Increment(ref _concurrentCount);
+
+            try
+            {
+                if (count != 1)
+                {
+                    throw new CryptographicException(SR.Cryptography_ConcurrentUseNotSupported);
+                }
+
+                SafeBCryptHashHandle clone = Interop.BCrypt.BCryptDuplicateHash(_hashHandle);
+                return new Shake256(clone);
+            }
+            finally
+            {
+                Interlocked.Decrement(ref _concurrentCount);
+            }
+        }
+
+        public void Dispose()
+        {
+            _hashHandle?.Dispose();
+        }
+
+        private static SafeBCryptHashHandle CreateHashHandle()
+        {
+            NTSTATUS ntStatus = Interop.BCrypt.BCryptCreateHash(
+                (nuint)Interop.BCrypt.BCryptAlgPseudoHandle.BCRYPT_CSHAKE256_ALG_HANDLE,
+                out SafeBCryptHashHandle hashHandle,
+                pbHashObject: IntPtr.Zero,
+                cbHashObject: 0,
+                secret: ReadOnlySpan<byte>.Empty,
+                cbSecret: 0,
+                Interop.BCrypt.BCryptCreateHashFlags.None);
+
+            if (ntStatus != NTSTATUS.STATUS_SUCCESS)
+            {
+                hashHandle.Dispose();
+                throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+            }
+
+            return hashHandle;
+        }
+
+        private void Act(Action action)
+        {
+            if (Interlocked.Increment(ref _concurrentCount) == 1)
+            {
+                try
+                {
+                    action();
+                }
+                finally
+                {
+                    Interlocked.Decrement(ref _concurrentCount);
+                }
+            }
+            else
+            {
+                throw new InvalidOperationException("Concurrent access is not allowed.");
+            }
+        }
+#else // !NETFRAMEWORK
+#pragma warning disable CA1822
+#pragma warning disable IDE0060
+
+        // The .NET Standard 2.0 and 2.1 builds just need the shape to exist.
+        // Shake256 exists in .NET 8 and later.
+
+        internal Shake256()
+        {
+            throw new PlatformNotSupportedException();
+        }
+
         internal void AppendData(ReadOnlySpan<byte> data)
         {
-            throw new NotImplementedException();
         }
 
         internal void GetHashAndReset(Span<byte> destination)
         {
-            throw new NotImplementedException();
         }
 
         internal void GetCurrentHash(Span<byte> destination)
         {
-            throw new NotImplementedException();
         }
 
         internal Shake256 Clone()
         {
-            throw new NotImplementedException();
+            return null!;
         }
 
         public void Dispose()
         {
         }
+#pragma warning disable IDE0060
+#pragma warning restore CA1822
+#endif
     }
 #elif NET8_0
     internal static class Shake256Extensions
```

Comment:
Why is this an extension instead of just a conditional inside the actual `Shake256` class?